File: capacity-club-api/src/domain-modules/home-training/model/entity/home-training.entity.ts
import { uniqueIdTransformer, UniqueId, BaseEntity } from '@common/model';
import { ApiProperty } from '@nestjs/swagger';
import { Transform } from 'class-transformer';
import { MemberHomeTraining } from 'domain-modules/member-home-training/model/entity/member_home_training.entity';
import { Workout } from 'domain-modules/workout/model/entity/workout.entity';
import { Entity, PrimaryColumn, Column, OneToMany } from 'typeorm';

/**
 * Entity representing a home training program.
 * A home training consists of multiple workouts and contains details about the training schedule.
 * It is a training program that the member performs from home or any other place of their choice.
 */
@Entity()
export class HomeTraining extends BaseEntity {
  /**
   * Unique identifier for the home training program.
   * Uses a custom transformer for serialization.
   */
  @PrimaryColumn('varchar')
  @Transform(uniqueIdTransformer.to, { toClassOnly: true })
  @Transform(uniqueIdTransformer.from, { toPlainOnly: true })
  home_training_id: UniqueId;
  @Column({ nullable: false })
  title: string;
  @Column({ nullable: true })
  description: string;
  @Column({ nullable: false })
  nb_week: number;
  @Column({ nullable: false })
  nb_training_by_week: number;
  @Column({ nullable: false })
  price: number;

  @OneToMany(() => Workout, (workout) => workout.home_training, {
    cascade: true,
    lazy: true,
  })
  @ApiProperty({ type: () => Workout, isArray: true })
  workouts: Promise<Workout[]>;

  @OneToMany(
    () => MemberHomeTraining,
    (member_home_training) => member_home_training.home_training,
    {
      cascade: true,
      lazy: true,
    },
  )
  @ApiProperty({ type: () => MemberHomeTraining, isArray: true })
  member_home_trainings: Promise<MemberHomeTraining[]>;
}

File: capacity_club_mobile_app/lib/domain-features/domain/home-training/entity/home_training_entity.dart
import 'package:capacity_club_mobile_app/core/model/entities/unique_id.dart';
import 'package:capacity_club_mobile_app/domain-features/domain/member-home-training/entity/member_home_training_entity.dart';
import 'package:capacity_club_mobile_app/domain-features/domain/workout/entity/workout_entity.dart';

class HomeTrainingEntity {
  final UniqueId home_training_id;
  final String title;
  final String description;
  final int nb_week;
  final int nb_training_by_week;
  final double price;
  final List<WorkoutEntity> workouts;
  final MemberHomeTrainingEntity member_home_trainings;
  final bool is_empty;

  HomeTrainingEntity(
      {required this.home_training_id,
      required this.title,
      required this.description,
      required this.nb_week,
      required this.nb_training_by_week,
      required this.price,
      required this.workouts,
      required this.member_home_trainings,
      required this.is_empty});

  factory HomeTrainingEntity.empty() {
    return HomeTrainingEntity(
        home_training_id: UniqueId(''),
        title: '',
        description: '',
        nb_week: 0,
        nb_training_by_week: 0,
        price: 0,
        workouts: [],
        member_home_trainings: MemberHomeTrainingEntity.empty(),
        is_empty: true);
  }
}
